internal/backend/runtime/omni/controllers/omni/machine_config_gen_options.go

```diff
@@ -79,18 +83,44 @@ func GenInstallConfig(machineStatus *omni.MachineStatus, clusterMachineTalosVers
 
 	installDisk := omni.GetMachineStatusSystemDisk(machineStatus)
 
-	diskSize := ^uint64(0)
+	var selected *specs.MachineStatusSpec_HardwareStatus_BlockDevice
+
+	const transportUSB = "usb"
+
+	shouldSelect := func(other *specs.MachineStatusSpec_HardwareStatus_BlockDevice) bool {
+		if selected == nil {
+			return true
+		}
+
+		if other.Transport == transportUSB && selected.Transport != transportUSB {
+			return false
+		}
+
+		if other.Transport != transportUSB && selected.Transport == transportUSB {
+			return true
+		}
+
+		if other.Size < selected.Size {
+			return true
+		}
+
+		return false
+	}
 
 	if installDisk == "" {
 		for _, disk := range machineStatus.TypedSpec().Value.Hardware.Blockdevices {
 			if disk.Readonly || disk.Type == storage.Disk_CD.String() {
```

Comment:
I would rewrite this into two statements which are both testable and (I think) easier to understand:

```
candidates := allBlockDevices()

// absolute filtering
candidates := Filter(candidates, !read_only, !CD, size > min_disk_size)

candidates := Sort(candidates, prefer !USB, sort by size asc)

pickedDisk := candidates[0]
```